pkg/matcher/matcher_test.go

```diff
@@ -183,3 +183,19 @@ another`
 	assert.False(t, r.Success)
 	assert.Equal(t, diff, r.Diff)
 }
+
+func TestFileMatcher_Validate(t *testing.T) {
+	m := FileMatcher{}
+	got := m.Match("zoom", "zoom.txt")
```

Comment:
I thought about the temp file approach but I wanted input since that didn't feel great either. Ordinarily I'd probably accept a `Reader` instead of a filename but then we're back to the original question of where does the file actually get opened.

Reply:
Right, the only other thing I can really think of is to `FileMatcher` take an argument for a function, that basically wraps ioutil.ReadFile(filePath string). Like:
```golang
type TestingMocking struct {
	readFile func(filePath string) ([]byte, error)
}
```
You could then setup `FileMatcher` in `NewMatcher`.
And then:
```golang
func (m FileMatcher) Match(got interface{}, expected interface{}) MatcherResult {
	buf, err := m.readFile(expected.(string))
        // rest goes here 
```
This would allow us to mock readFile. Again I think this may be too robust. I think I'll defer @SimonBaeumer on this as I don't really have preference.

Reply:
Good call, I'll try that out and see how it ends up.

Reply:
Hey Together,
so yes, ideally we would mock the filesystem but this not often done and not necessary imho.
I am totally fine with using libraries and access to filesystem or env variables.

The downside of mocking is in general that it gets very complex and often takes a lot of effort to maintain.
Mocking is nice if there is some complex things I want to replace, like network connections, tasks that need a lot of time or simply complex stuff to setup, i.e. you would need to run a kubernetes cluster.

Go ignores directories starting with underscores, like `_assets` and directories matching `testdata`. 
Just put the asset inside a `pkg/matcher/testdata` directory.
You should be able access it like this:
```
wd, _ := os.Getwd()
asset := path.Join(wd, "testdata/file.txt")
content, _ := ioutil.ReadFile(asset)
```

The testing framework of go will execute all tests with the working directory pointing to the currently executed file, so it is easy to access the testdata.

**Mocking:**

Another option like @dylanhitt mentioned is to use dependency injection. That means that we inject (normally as input parameters in the constructor or functions) dependencies on other structs (or classes). 
These classes / structs do implement an interface, therefore you could replace the implementation inside your tests with a fake one.

Original implementation, i.e. in `file_matcher.go`:

```
type FileReaderInterface interface  {
    func Read(file string) string
}
type FileReader struct {}
func (r *FileReader) Read (file strinng) string {
    ioutil.ReadFile...
    // direct access to filesystem
}

type FileMatcher struct {
    Reader FileReaderInterface
}
func NewFileMatcher() *FileMatcher {
    var reader FileReaderInterface
    reader = FileReader{}
    return &FileMatcher{
        Reader: reader,
    }
}
```

In your test `file_matcher_test.go` you would than do something like this:

```
type FileReaderMock {}
func Read(file string) string {
    if file == "test.txt" {
        return "Mocked this for a test"
    }

    if file == "test2.txt" {
       return "mocked for another test"
    }
    // maybe panic if the file wasn't mocked
    panic("file was not mocked")
}

func TestFileMatcher(t *testing.T) {
    var readerMock FileReaderInterface
    readerMock = FileReaderMock{}

    // Inject the reader mock into the FileMatcher
    matcher := FileMatcher{Reader: readerMock}
} 
```